### ISSUE TITLE
Accessing Particular Nightly Builds Don't Work

### ISSUE THREAD

**Issue by anthonyalayo** [pytorch/pytorch#110250]
### 🐛 Describe the bug

I would like to use a particular nightly release version of libtorch. I need to do this because:

1. Official version releases are sparse
2. Relying on "latest" is dangerous -- consistent builds/deployments is much safer

On the official page:
https://pytorch.org/get-started/locally/

I can only see the official releases or the nightly releases:
![image](https://github.com/pytorch/pytorch/assets/17555572/d3f36dcb-5767-4dee-b737-7834e13f9798)

However, in the github workflows, I can see that date stamped releases are being uploaded as well. For example, with this run flavor:
https://github.com/pytorch/pytorch/actions/runs/6335907738/job/17207923080

I can see:
```
record_file=libtorch-cxx11-abi-shared-with-deps-2.2.0.dev20230928+cpu.zip
..
record_file=libtorch-cxx11-abi-shared-with-deps-latest.zip
```

However, attempting to access these urls results in an 404. If I attempt a similar URL with the wheel storage:
https://download.pytorch.org/whl/nightly/cpu/torch-2.1.0.dev20230901-cp310-none-macosx_11_0_arm64.whl

It works as expected. Can this be fixed?

Thanks,
Anthony

### Versions

N/A

**Comment by malfet** [comment]
Aren't artifacts publicly available from the workflow page for  30 days or so:
https://github.com/pytorch/pytorch/actions/runs/6335907738
<img width="770" alt="image" src="https://github.com/pytorch/pytorch/assets/2453524/3a2e226a-09fc-4905-9fa8-f1477720c7d7">


And I don't think we still do arm64 libtroch binaries (Though if you have a PR I will gladly review it)
Also, for MacOS platform there are no difference between libtorch libraries and `libtorch.dylib` bundled into the wheel or the conda package, so one can simply extract it from there)

**Comment by malfet** [comment]
But it sounds reasonable, that https://download.pytorch.org/libtorch/ should be indexable

**Comment by anthonyalayo** [comment]
@malfet Right, I'm referring to https://download.pytorch.org/libtorch/ links. I have a simple script that downloads the nightly using the date/version, and that isn't working with the nightly releases for libtorch. Why is that?

**Comment by anthonyalayo** [comment]
@malfet any thoughts on that last comment?

**Comment by malfet** [comment]
@anthonyalayo mind sharing a reference to the script. Here is the content of  `s3://pytorch/libtorch/nightly/cpu/`:
```shell
% aws s3 ls s3://pytorch/libtorch/nightly/cpu/libtorch-mac
...
2023-09-11 13:53:23  175551575 libtorch-macos-2.2.0.dev20230911.zip
2023-09-12 02:29:20  175657404 libtorch-macos-2.2.0.dev20230912.zip
2023-09-13 02:26:49  175679963 libtorch-macos-2.2.0.dev20230913.zip
2023-09-14 02:30:02  175681091 libtorch-macos-2.2.0.dev20230914.zip
2023-09-15 02:27:05  175693946 libtorch-macos-2.2.0.dev20230915.zip
2023-09-16 02:24:22  175700853 libtorch-macos-2.2.0.dev20230916.zip
2023-09-17 02:25:21  175701711 libtorch-macos-2.2.0.dev20230917.zip
2023-09-18 02:27:23  175695032 libtorch-macos-2.2.0.dev20230918.zip
2023-09-19 11:16:26  175781919 libtorch-macos-2.2.0.dev20230919.zip
2023-09-20 02:26:04  175718209 libtorch-macos-2.2.0.dev20230920.zip
2023-09-21 02:27:43  175744315 libtorch-macos-2.2.0.dev20230921.zip
2023-09-22 02:26:00  175748271 libtorch-macos-2.2.0.dev20230922.zip
2023-09-23 02:27:29  175773726 libtorch-macos-2.2.0.dev20230923.zip
2023-09-24 02:28:08  175776950 libtorch-macos-2.2.0.dev20230924.zip
2023-09-25 02:30:46  175820111 libtorch-macos-2.2.0.dev20230925.zip
2023-09-26 02:59:04  175818953 libtorch-macos-2.2.0.dev20230926.zip
2023-09-27 02:26:24  175845824 libtorch-macos-2.2.0.dev20230927.zip
2023-09-28 02:30:10  175870139 libtorch-macos-2.2.0.dev20230928.zip
2023-09-29 02:27:05  175883529 libtorch-macos-2.2.0.dev20230929.zip
2023-09-30 02:26:03  175876692 libtorch-macos-2.2.0.dev20230930.zip
2023-10-01 02:27:33  175876896 libtorch-macos-2.2.0.dev20231001.zip
2023-10-02 02:50:40  175876862 libtorch-macos-2.2.0.dev20231002.zip
2023-10-03 02:32:28  175883167 libtorch-macos-2.2.0.dev20231003.zip
2023-10-04 02:27:04  175786886 libtorch-macos-2.2.0.dev20231004.zip
2023-10-05 02:28:05  175859064 libtorch-macos-2.2.0.dev20231005.zip
2023-10-06 02:29:21  175869555 libtorch-macos-2.2.0.dev20231006.zip
2023-10-06 02:29:23  175869555 libtorch-macos-latest.zip
...
```

As I've mentioned, there are currently no libtorch builds for Apple Silicon, if you are willing to propose a PR that folds 2 binary builds jobs into one (as there are no difference between libtorch builds and wheels builds for MacOS platform) I'll be more than happy to review it

**Comment by anthonyalayo** [comment]
@malfet 

> Here is the content of s3://pytorch/libtorch/nightly/cpu/

Right, that's exactly what I would like to be indexable. Right now https://download.pytorch.org/libtorch returns a 403.

Is it possible to make that public, the same as https://download.pytorch.org/whl/?

> As I've mentioned, there are currently no libtorch builds for Apple Silicon, if you are willing to propose a PR that folds 2 binary builds jobs into one (as there are no difference between libtorch builds and wheels builds for MacOS platform) I'll be more than happy to review it

I'm willing if I can get a pointer in the right direction.

> mind sharing a reference to the script.

```
import platform
import os
import urllib.request

PYTORCH_VERSION = "2.1.0"

if platform.system() == "Darwin":
    PYTORCH_MODE = "cpu"
    PYTORCH_PYTHON_VERSION = "cp310-none"

    if platform.machine() == "x86_64":
        PYTORCH_PLATFORM = "macosx_10_9_x86_64"
    else:
        PYTORCH_PLATFORM = "macosx_11_0_arm64"

else:
    PYTORCH_PYTHON_VERSION = "cp310-cp310"

    CUDAToolkit_ROOT = "/usr/local/cuda"
    os.environ["CUDA_PATH"] = CUDAToolkit_ROOT

    if os.path.exists(CUDAToolkit_ROOT):
        PYTORCH_MODE = "cu118"
    else:
        PYTORCH_MODE = "cpu"

    if platform.machine() == "x86_64":
        PYTORCH_PLATFORM = "linux_x86_64"
    elif platform.machine() == "aarch64":
        PYTORCH_MODE = "cpu"
        PYTORCH_PLATFORM = "manylinux_2_17_aarch64.manylinux2014_aarch64"
    else:
        raise Exception(f"No PyTorch support for architecture {platform.machine()}")

PYTORCH_WHEEL_NAME = f"torch-{PYTORCH_VERSION}-{PYTORCH_PYTHON_VERSION}-{PYTORCH_PLATFORM}.whl"
PYTORCH_WHEEL_URL = f"https://download.pytorch.org/whl/{PYTORCH_MODE}/{PYTORCH_WHEEL_NAME}"

print(f"Fetching PyTorch binary distribution from {PYTORCH_WHEEL_URL}")
urllib.request.urlretrieve(PYTORCH_WHEEL_URL, PYTORCH_WHEEL_NAME)
```

**Comment by anthonyalayo** [comment]
@malfet any thoughts here? can we open the gate? :-)

**Comment by anthonyalayo** [comment]
@malfet bump in case this got forgotten

**Comment by malfet** [comment]
@anthonyalayo https://download.pytorch.org/libtorch/nightly/cpu should now be indexable after https://github.com/pytorch/builder/commit/13c071f2783ec4100fae94c74d2285ccedf5acb7

As for the static builds, I've just disabled them as part of https://github.com/pytorch/pytorch/pull/112452 because they were never published in getting started nor really ever tested

**Comment by anthonyalayo** [comment]
@malfet sad news about the static builds, but good news overall. Thank you! I will test this out now.

**Comment by malfet** [comment]
@anthonyalayo can you explain the usecase for a static builds? (As CUDA static builds has been essential a non-option after CUDA-11.4 that resulted in 2+Gb that could not be linked with 32-bit relptx from cubins, maintaining the rest felt quite weird)

**Comment by anthonyalayo** [comment]
@malfet 

I just tried out https://download.pytorch.org/libtorch/nightly/cpu, but unfortunately every link is resulting in:

```
<Error>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>Z4Q292KXW371Y8NS</RequestId>
<HostId>8s6ncud3sULMmgfwGw38ttaC8st2Cbfz3dSJU7+LtgPSQFa6Jo2/FlUa+mVCiXfuQOm0q+A9rA8=</HostId>
</Error>
```

Could you apply the permissions fix to the files inside as well?

> can you explain the usecase for a static build?

I'll have to really profile both to give you a proper answer, but the idea was faster performance in theory.

**Comment by anthonyalayo** [comment]
Also, I can't seem to find these

https://github.com/pytorch/pytorch/actions/workflows/generated-macos-arm64-binary-libtorch-cxx11-abi-nightly.yml

![image](https://github.com/pytorch/pytorch/assets/17555572/9d84aabb-1b49-4d47-8ec0-08758eab56b1)

In which folder would macOS arm64 binaries be?

**Comment by anthonyalayo** [comment]
@malfet this is still a problem, could you please fix the permission issue?

Thanks!

**Comment by anthonyalayo** [comment]
Friendly ping to other power users, @ezyang @zou3519

**Comment by anthonyalayo** [comment]
Polite ping @malfet @ezyang @zou3519

**Comment by malfet** [comment]
@anthonyalayo why don't want you to provide a PR to fix this one?

**Comment by anthonyalayo** [comment]
> @anthonyalayo why don't want you to provide a PR to fix this one?

I can provide a PR for this? I had no idea, where in the code is it maintained? I can check.

**Comment by anthonyalayo** [comment]
pinging back here @malfet
